src/Extension.Tests/CodeRunEvaluationTests.cs

```diff
@@ -0,0 +1,290 @@
+﻿using System;
+using System.Threading.Tasks;
+using FluentAssertions;
+using Xunit;
+using Microsoft.DotNet.Interactive;
+using Microsoft.DotNet.Interactive.CSharp;
+using Extension.Tests.Utilities;
+using Extension.Events;
+
+namespace Extension.Tests
+{
+    public class CodeRunEvaluationTests
+    {
+        [Fact]
+        public async Task submitting_code_that_fails_evaluation_criterion_gives_failed_evaluation()
+        {
+            var evaluator = new Evaluator();       
+            using var csharpKernel = new CSharpKernel();
+            using var kernel = new CompositeKernel { csharpKernel };
+            kernel
+                .UseQuestionMagicCommand(evaluator)
+                .UseAnswerMagicCommand(evaluator);
+
+
+            using var events = kernel.KernelEvents.ToSubscribedList();
+
+            // instructor provides testcase script which returns bool
+            await kernel.SubmitCodeAsync(@"
+#!answer 5
+var testcases = new (double, double, double)[]
+{
+    (0, 0, 0),
+    (0.5, 1.0, 0.25),
+    (100, 50, 2500)
+};
+
+testcases.All(testcase => triangleArea(testcase.Item1, testcase.Item2) == testcase.Item3)
+");
+
+            // students code runs
+            await kernel.SubmitCodeAsync(@"
+#!question 5
+
+double triangleArea(double x, double y) 
+{
+    return x * y;
+}
+");
+
+            events.Should().ContainSingle<CodeRunEvaluationProduced>()
+                .Which.Evaluation.Passed.Should().Be(false);
+        }
+
+        [Fact]
+        public async Task submitting_code_that_passes_evaluation_criterion_gives_passed_evaluation()
+        {
+            var evaluator = new Evaluator();
+            using var csharpKernel = new CSharpKernel();
+            using var kernel = new CompositeKernel { csharpKernel };
+            kernel
+                .UseQuestionMagicCommand(evaluator)
+                .UseAnswerMagicCommand(evaluator);
+
+
+            using var events = kernel.KernelEvents.ToSubscribedList();
+
+            // instructor provides testcase script which returns bool
+            await kernel.SubmitCodeAsync(@"
+#!answer 6
+var testcases = new (double, double, double)[]
+{
+    (0, 0, 0),
+    (0.5, 1.0, 0.25),
+    (100, 50, 2500)
+};
+
+testcases.All(testcase => triangleArea(testcase.Item1, testcase.Item2) == testcase.Item3)
+");
```

Comment:
These code submissions are fairly verbose and look to be either the same from one test to another or substantially similar.

A test can communicate its intent a lot more clearly if you don't make the person reading it have to infer the intent by reading this much code. One tactic for making the test more concise is to hide the details that don't matter, and let people dig into them as needed. For example, extract these code submissions into fields with names that signal their intent.

Reply:
Tests have all been refactored